File: controllers/leaveController.js
// =====================================================
// Controller: Đơn nghỉ phép
// =====================================================
const LeaveModel = require('../models/leaveModel');

const getAll = async (req, res) => {
    try {
        const options = {
            page: parseInt(req.query.page) || 1,
            limit: parseInt(req.query.limit) || 20,
            trangThai: req.query.trangThai || '',
            search: req.query.search || '',
            maNhanVien: req.user.tenQuyen === 'NhanVien' ? req.user.maNhanVien : (req.query.maNhanVien || '')
        };

        const [rows, total] = await Promise.all([
            LeaveModel.getAll(options),
            LeaveModel.count(options)
        ]);

        return res.json({
            success: true, data: rows,
            pagination: { page: options.page, limit: options.limit, total, totalPages: Math.ceil(total / options.limit) }
        });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi lấy danh sách nghỉ phép' });
    }
};

const getOne = async (req, res) => {
    try {
        const leave = await LeaveModel.findById(req.params.id);
        if (!leave) return res.status(404).json({ success: false, message: 'Không tìm thấy đơn' });
        return res.json({ success: true, data: leave });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi máy chủ' });
    }
};

const create = async (req, res) => {
    try {
        const data = {
            ...req.body,
            maNhanVien: req.user.tenQuyen === 'NhanVien' ? req.user.maNhanVien : req.body.maNhanVien
        };
        const id = await LeaveModel.create(data);
        const leave = await LeaveModel.findById(id);
        return res.status(201).json({ success: true, message: 'Tạo đơn nghỉ phép thành công', data: leave });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi tạo đơn' });
    }
};

const update = async (req, res) => {
    try {
        await LeaveModel.update(req.params.id, req.body);
        const leave = await LeaveModel.findById(req.params.id);
        return res.json({ success: true, message: 'Cập nhật thành công', data: leave });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi cập nhật' });
    }
};

const approve = async (req, res) => {
    try {
        await LeaveModel.approve(req.params.id);
        return res.json({ success: true, message: 'Duyệt đơn nghỉ phép thành công' });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi duyệt đơn' });
    }
};

const reject = async (req, res) => {
    try {
        await LeaveModel.reject(req.params.id);
        return res.json({ success: true, message: 'Từ chối đơn nghỉ phép' });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi từ chối đơn' });
    }
};

const remove = async (req, res) => {
    try {
        await LeaveModel.delete(req.params.id);
        return res.json({ success: true, message: 'Xóa đơn thành công' });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi xóa đơn' });
    }
};

module.exports = { getAll, getOne, create, update, approve, reject, remove };

File: quan.comdemo/quan.comdemo/src/main/java/K23/CNT1/_5/DAQ/quan/comdemo/controller/LeaveController.java
package K23.CNT1._5.DAQ.quan.comdemo.controller;

import io.jsonwebtoken.Claims;
import K23.CNT1._5.DAQ.quan.comdemo.dto.ApiResponse;
import K23.CNT1._5.DAQ.quan.comdemo.entity.DonNghiPhep;
import K23.CNT1._5.DAQ.quan.comdemo.service.LeaveService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/leaves")
public class LeaveController {

    @Autowired
    private LeaveService leaveService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<DonNghiPhep>>> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit,
            @RequestParam(required = false) Integer maNhanVien,
            @RequestParam(required = false, defaultValue = "") String trangThai,
            @RequestParam(required = false, defaultValue = "") String search) {

        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth != null && auth.getPrincipal() instanceof Claims claims) {
            String tenQuyen = claims.get("tenQuyen", String.class);
            if ("NhanVien".equals(tenQuyen)) {
                maNhanVien = claims.get("maNhanVien", Integer.class);
            }
        }

        Page<DonNghiPhep> resultPage = leaveService.getAll(page, limit, maNhanVien, trangThai, search);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("page", page);
        pagination.put("limit", limit);
        pagination.put("total", resultPage.getTotalElements());
        pagination.put("totalPages", resultPage.getTotalPages());

        return ResponseEntity.ok(ApiResponse.ok(resultPage.getContent(), pagination));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<DonNghiPhep>> getOne(@PathVariable Integer id) {
        DonNghiPhep dnp = leaveService.getOne(id);
        if (dnp == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("Không tìm thấy đơn"));
        }
        return ResponseEntity.ok(ApiResponse.ok(dnp));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<DonNghiPhep>> create(@RequestBody Map<String, Object> body) {
        try {
            Integer userMaNV = null;
            String userRole = null;

            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            if (auth != null && auth.getPrincipal() instanceof Claims claims) {
                userRole = claims.get("tenQuyen", String.class);
                userMaNV = claims.get("maNhanVien", Integer.class);
            }

            DonNghiPhep created = leaveService.create(body, userMaNV, userRole);
            return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok("Tạo đơn nghỉ phép thành công", created));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi tạo đơn: " + e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<DonNghiPhep>> update(@PathVariable Integer id, @RequestBody Map<String, Object> body) {
        try {
            DonNghiPhep updated = leaveService.update(id, body);
            return ResponseEntity.ok(ApiResponse.ok("Cập nhật thành công", updated));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi cập nhật: " + e.getMessage()));
        }
    }

    @PutMapping("/{id}/approve")
    public ResponseEntity<ApiResponse<DonNghiPhep>> approve(@PathVariable Integer id) {
        try {
            DonNghiPhep approved = leaveService.approve(id);
            return ResponseEntity.ok(ApiResponse.ok("Duyệt đơn nghỉ phép thành công", approved));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi duyệt đơn: " + e.getMessage()));
        }
    }

    @PutMapping("/{id}/reject")
    public ResponseEntity<ApiResponse<DonNghiPhep>> reject(@PathVariable Integer id) {
        try {
            DonNghiPhep rejected = leaveService.reject(id);
            return ResponseEntity.ok(ApiResponse.ok("Từ chối đơn nghỉ phép", rejected));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi từ chối đơn: " + e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<String>> delete(@PathVariable Integer id) {
        try {
            leaveService.delete(id);
            return ResponseEntity.ok(ApiResponse.ok("Xóa đơn thành công", null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi xóa đơn: " + e.getMessage()));
        }
    }
}
